ddtrace/internal/datadog/profiling/dd_wrapper/src/profile.cpp

```diff
@@ -0,0 +1,228 @@
+#include "profile.hpp"
+#include "libdatadog_helpers.hpp"
+
+#include <functional>
+#include <iostream>
+
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+
+// Inline helpers
+namespace {
+
+inline bool
+make_profile(const ddog_prof_Slice_ValueType& sample_types,
+             const struct ddog_prof_Period* period,
+             ddog_prof_Profile& profile)
+{
+    ddog_prof_Profile_NewResult res = ddog_prof_Profile_new(sample_types, period, nullptr);
+    if (res.tag != DDOG_PROF_PROFILE_NEW_RESULT_OK) {
+        const std::string errmsg = Datadog::err_to_msg(&res.err, "Error initializing profile");
+        std::cerr << errmsg << std::endl;
+        ddog_Error_drop(&res.err);
+        return false;
+    }
+    profile = res.ok;
+    return true;
+}
+
+}
+
+using namespace Datadog;
+
+void
+Profile::reset()
+{
+    // Drop the profiles
+    if (cur_profile.inner != nullptr) {
+        ddog_prof_Profile_drop(&cur_profile);
+    }
+    if (last_profile.inner != nullptr) {
+        ddog_prof_Profile_drop(&last_profile);
+    }
+
+    // non-RAII heap storage has been dropped, state has been reinitialized
+    // we can pretend it's the first time all over again
+    first_time.store(true);
+}
+
+bool
+Profile::cycle_buffers()
+{
+    const std::lock_guard<std::mutex> lock(profile_mtx);
+    std::swap(last_profile, cur_profile);
```

Comment:
several threads can update the cur_profile ?

Reply:
Indeed, extended the protection.  In particular, there's one exemplary case where the critical period is defined by the caller.  In that instance (uploading), I added the concept of `borrowing` and then `releasing` the profile, which uses the same lock as the other internal function.

I think this is better now, PTAL